crates/autopilot/src/solvable_orders.rs

```diff
@@ -348,6 +313,66 @@ impl SolvableOrdersCache {
         Ok(())
     }
 
+    async fn fetch_balances(&self, queries: Vec<Query>) -> HashMap<Query, U256> {
+        let fetched_balances = {
+            let _timer = self.stage_timer("balance_fetch");
+            self.balance_fetcher.get_balances(&queries).await
+        };
+        queries
+            .into_iter()
+            .zip(fetched_balances)
+            .filter_map(|(query, balance)| match balance {
+                Ok(balance) => Some((query, balance)),
+                Err(err) => {
+                    tracing::warn!(
+                        owner = ?query.owner,
+                        token = ?query.token,
+                        source = ?query.source,
+                        error = ?err,
+                        "failed to get balance"
+                    );
+                    None
+                }
+            })
+            .collect()
+    }
+
+    /// Executed effectful orders filtering in parallel.
+    async fn filter_invalid_orders(
+        &self,
+        mut orders: Vec<Order>,
+        counter: &mut OrderFilterCounter,
+        invalid_order_uids: &mut HashSet<OrderUid>,
+    ) -> Vec<Order> {
+        let banned_user_orders_fut = async {
+            let _timer = self.stage_timer("banned_user_filtering");
+            find_banned_user_orders(&orders, &self.banned_users).await
+        };
+        let invalid_signature_orders_fut = async {
+            let _timer = self.stage_timer("invalid_signature_filtering");
+            find_invalid_signature_orders(&orders, self.signature_validator.as_ref()).await
+        };
+        let unsupported_token_orders_fut = async {
+            let _timer = self.stage_timer("unsupported_token_filtering");
+            find_unsupported_tokens(&orders, self.bad_token_detector.clone()).await
+        };
+        let (banned_user_orders, invalid_signature_orders, unsupported_token_orders) = tokio::join!(
+            banned_user_orders_fut,
+            invalid_signature_orders_fut,
+            unsupported_token_orders_fut,
+        );
+
+        counter.checkpoint_by_invalid_orders("banned_user", &banned_user_orders);
```

Comment:
These can double count orders since a single order could get filtered out for multiple reasons. But I think this is fine to not open another can of worms before merging this.

Reply:
Good catch, thanks. Used another counter for those.

Reply:
Sweet, your fix is actually super simple. The results of `.remove()` came in handy after all. 👍